### DISCUSSION TITLE
does cast storage need contract abi?

### DISCUSSION THREAD

**Discussion by jaimebarrancos** [Cyfrin/security-and-auditing-full-course-s23#95] [Q&A] [answered]
While trying to solve challenge of section 3, when I run cast storage it always tries to get the code, searching my local files, and then searching Etherscan for the contracts. 
I thought having the code of the contracts was not required for the command to work.

Here is what happens when it tries to fetch the code from Etherscan:
```bash
No matching artifacts found, fetching source code from Etherscan...

Error: 
Compiler run failed:
Error (6275): Source "lib/ctf/src/protocol/Challenge.sol" not found: File not found. Searched the following locations: "/private/var/folders/70/pyz3jhgd6ps9mb2qgll4g50r0000gn/T/.tmp17faC3".
 --> S3/src/challenges/S3.sol:4:1:
  |
4 | import {Challenge} from "lib/ctf/src/protocol/Challenge.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

When running:
```
cast storage --rpc-url=http://127.0.0.1:8545 0xA2626bE06C11211A44fb6cA324A67EBDBCd30B70
```

I have a local anvil chain with an alchemy key to Sepolia setup.

**Comment by johnsonstephan** [comment, marked as answer]
**`cast storage`**

- **Use**: Get the raw value of a contract's storage slot or its full storage layout. `cast storage` [options] address slot
- **Description**: Get the raw value of a contract's storage slot. (Slot locations greater than 18446744073709551615 (u64::MAX) should be given as hex. Use cast index to compute mapping slots.) Emit the slot number to get the full storage layout (requires contract to be verified on Etherscan with a Solidity version > 0.6.5 or you must be in a Forge project with a local contract matching the deployed bytecode). The address (address) can be an ENS name or an address.

Source: [Foundry Book](https://book.getfoundry.sh/reference/cast/cast-storage)

**Reply by jaimebarrancos** [reply]
I've played around with it and, to use cast storage <slot> the ABI is not needed. But when not specifying the slot, the contract IS required . Why ?

I believe it's because the output would become unreadable / useless so it would only make sense to use cast storage with access to the code. This way foundry would return a usable version of the storage.

**Reply by johnsonstephan** [reply]
Take your time with the documentation as you are learning. 

--- 
**Examples**
1. Get the value of slot 0 on the WETH contract.
`cast storage 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2 0`
2. Get the full storage layout of the Milady contract. (this reflects your case of "But when not specifying the slot, the contract IS required")
`cast storage 0x5Af0D9827E0c53E4799BB226655A1de152A425a5`

As you're saying, it would be "useless" without the contract. 

---

Suggest using the `--help` flag as you're learning as well.

_Same Source_
